src/Go/boardAnalysis/patternMatching.ts

```diff
@@ -0,0 +1,187 @@
+// Inspired by https://github.com/pasky/michi/blob/master/michi.py
+import { BoardState, PlayerColor, playerColors, PointState } from "../boardState/goConstants";
+import { sleep } from "./goAI";
+
+export const threeByThreePatterns = [
+  // 3x3 piece patterns; X,O are color pieces; x,o are any state except the opposite color piece;
+  // " " is off the edge of the board; "?" is any state (even off the board)
+  [
+    "XOX", // hane pattern - enclosing hane
+    "...",
+    "???",
+  ],
+  [
+    "XO.", // hane pattern - non-cutting hane
+    "...",
+    "?.?",
+  ],
+  [
+    "XO?", // hane pattern - magari
+    "X..",
+    "o.?",
+  ],
+  [
+    ".O.", // generic pattern - katatsuke or diagonal attachment; similar to magari
+    "X..",
+    "...",
+  ],
+  [
+    "XO?", // cut1 pattern (kiri] - unprotected cut
+    "O.x",
+    "?x?",
+  ],
+  [
+    "XO?", // cut1 pattern (kiri] - peeped cut
+    "O.X",
+    "???",
+  ],
+  [
+    "?X?", // cut2 pattern (de]
+    "O.O",
+    "xxx",
+  ],
+  [
+    "OX?", // cut keima
+    "x.O",
+    "???",
+  ],
+  [
+    "X.?", // side pattern - chase
+    "O.?",
+    "   ",
+  ],
+  [
+    "OX?", // side pattern - block side cut
+    "X.O",
+    "   ",
+  ],
+  [
+    "?X?", // side pattern - block side connection
+    "o.O",
+    "   ",
+  ],
+  [
+    "?XO", // side pattern - sagari
+    "o.o",
+    "   ",
+  ],
+  [
+    "?OX", // side pattern - cut
+    "X.O",
+    "   ",
+  ],
+];
+
+/**
+ * Searches the board for any point that matches the expanded pattern set
+ */
+export async function findAnyMatchedPatterns(
+  boardState: BoardState,
+  player: PlayerColor,
+  availableSpaces: PointState[],
+) {
+  const board = boardState.board;
+  const boardSize = board[0].length;
+  const patterns = expandAllThreeByThreePatterns();
+  for (let x = 0; x < boardSize; x++) {
+    for (let y = 0; y < boardSize; y++) {
+      const neighborhood = getNeighborhood(boardState, x, y);
+      const matchedPattern = patterns.find((pattern) => checkMatch(neighborhood, pattern, player));
+      if (
+        matchedPattern &&
+        availableSpaces.find((availablePoint) => availablePoint.x === x && availablePoint.y === y)
+      ) {
+        return board[x][y];
+      }
+    }
+    await sleep(10);
+  }
+  return null;
+}
+
+/**
+  Returns false if any point does not match the pattern, and true if it matches fully.
+ */
+function checkMatch(neighborhood: PointState[][], pattern: string[], player: PlayerColor) {
+  const patternArr = `${pattern[0]}${pattern[1]}${pattern[2]}`.split("");
+  const neighborhoodArray = [...neighborhood[0], ...neighborhood[1], ...neighborhood[2]];
+  const mismatch = patternArr.find((str, index) => !matches(str, neighborhoodArray[index], player));
+  return !mismatch;
```

Comment:
To avoid both negations, you could say instead:
```suggestion
  return patternArr.every((str, index) => matches(str, neighborhoodArray[index], player));
```